File: android/app/src/main/java/com/openburnbar/ui/calendar/CalendarPageLayout.kt
package com.openburnbar.ui.calendar

import android.content.Context
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import org.json.JSONArray
import org.json.JSONObject

// MARK: - Calendar Card Kind
//
// The registry of every analytics card the Calendar surface can render for
// the current day selection. Mirrors the macOS `CalendarCardKind`: each kind
// carries its own editorial metadata, and adding a card = add an entry here,
// prepare its data in `CalendarSelectionSnapshot`, and give it a renderer arm
// in `CalendarAnalyticsPanel`.

/** Grid span in columns (the analytics grid is 3 columns wide; 3 = full
 *  width). Surfaced in the UI as S/M/L sizes. */
enum class CalendarCardSpan(val columns: Int, val shortLabel: String) {
    S(1, "S"),
    M(2, "M"),
    L(3, "L"),
    ;

    fun next(): CalendarCardSpan = entries[(ordinal + 1) % entries.size]

    companion object {
        fun fromColumns(columns: Int?): CalendarCardSpan? = when (columns) {
            1 -> S
            2 -> M
            3 -> L
            else -> null
        }
    }
}

enum class CalendarCardKind(
    val key: String,
    val title: String,
    /** One-line "why it matters" — rendered as the card's footer microcopy. */
    val whyItMatters: String,
    val defaultSpan: CalendarCardSpan,
) {
    KPIS(
        "kpis",
        "Key Numbers",
        "The selection at a glance — spend, volume, sessions, cadence.",
        CalendarCardSpan.L,
    ),
    BURN_OVER_SELECTION(
        "burnOverSelection",
        "Burn Over Selection",
        "Day-by-day spend across the selected days.",
        CalendarCardSpan.L,
    ),
    PROVIDER_MIX(
        "providerMix",
        "Provider Mix",
        "Where the money actually goes across providers.",
        CalendarCardSpan.S,
    ),
    MODEL_MIX(
        "modelMix",
        "Model Mix",
        "Which models earn their keep — and which quietly dominate.",
        CalendarCardSpan.S,
    ),
    HOUR_OF_DAY_HEATMAP(
        "hourOfDayHeatmap",
        "When You Burn",
        "Your true working rhythm inside the selection.",
        CalendarCardSpan.L,
    ),
    PROJECT_FOCUS(
        "projectFocus",
        "Project Focus",
        "Which projects the selected days actually funded.",
        CalendarCardSpan.S,
    ),
    CACHE_ROI(
        "cacheROI",
        "Cache Savings",
        "Prompt caching pays rent. This is the receipt.",
        CalendarCardSpan.S,
    ),
    REASONING_SHARE(
        "reasoningShare",
        "Reasoning Share",
        "Extended thinking is a hidden line item. Watch its share.",
        CalendarCardSpan.S,
    ),
    ;

    companion object {
        fun fromKey(key: String?): CalendarCardKind? = entries.firstOrNull { it.key == key }
    }
}

// MARK: - Card configuration & page layout

/** One card's persisted state: which card, whether shown, and how wide. */
data class CalendarCardConfig(
    val kind: CalendarCardKind,
    val isVisible: Boolean = true,
    val span: CalendarCardSpan = kind.defaultSpan,
)

/**
 * The full panel layout: an ordered list of card configs, persisted as JSON
 * under [CalendarPageLayoutStore]. Same forward-compatible contract as the
 * macOS `CalendarPageLayout` — unknown kinds are dropped, missing kinds are
 * appended with defaults, so upgrades never clobber the user's arrangement.
 *
 * Instances are immutable; every mutation returns a new layout, which keeps
 * Compose state flow obvious (the screen holds one `MutableStateFlow`).
 */
data class CalendarPageLayout private constructor(val configs: List<CalendarCardConfig>) {

    /** Cards the grid actually renders, in order. */
    val visibleConfigs: List<CalendarCardConfig>
        get() = configs.filter { it.isVisible }

    val hiddenConfigs: List<CalendarCardConfig>
        get() = configs.filter { !it.isVisible }

    // MARK: Mutations

    /** Moves [kind] one slot earlier/later in the order. */
    fun move(kind: CalendarCardKind, delta: Int): CalendarPageLayout {
        val from = configs.indexOfFirst { it.kind == kind }
        if (from < 0) return this
        val to = (from + delta).coerceIn(0, configs.lastIndex)
        if (to == from) return this
        val next = configs.toMutableList()
        val config = next.removeAt(from)
        next.add(to, config)
        return CalendarPageLayout(next)
    }

    fun setVisible(kind: CalendarCardKind, visible: Boolean): CalendarPageLayout {
        val index = configs.indexOfFirst { it.kind == kind }
        if (index < 0) return this
        val next = configs.toMutableList()
        next[index] = next[index].copy(isVisible = visible)
        return CalendarPageLayout(next)
    }

    fun setSpan(kind: CalendarCardKind, span: CalendarCardSpan): CalendarPageLayout {
        val index = configs.indexOfFirst { it.kind == kind }
        if (index < 0) return this
        val next = configs.toMutableList()
        next[index] = next[index].copy(span = span)
        return CalendarPageLayout(next)
    }

    fun reset(): CalendarPageLayout = DEFAULT

    // MARK: JSON

    /** Compact JSON array: `[{"kind":"kpis","isVisible":true,"span":3}, …]`. */
    fun encode(): String {
        val array = JSONArray()
        for (config in configs) {
            array.put(
                JSONObject()
                    .put("kind", config.kind.key)
                    .put("isVisible", config.isVisible)
                    .put("span", config.span.columns),
            )
        }
        return array.toString()
    }

    companion object {
        val DEFAULT = CalendarPageLayout(CalendarCardKind.entries.map { CalendarCardConfig(kind = it) })

        /** JSON round-trip tolerant of unknown kinds and malformed payloads;
         *  anything unreadable yields [DEFAULT]. See type comment. */
        fun decode(json: String?): CalendarPageLayout {
            if (json.isNullOrBlank()) return DEFAULT
            val raw =
                try {
                    JSONArray(json)
                } catch (_: Exception) {
                    return DEFAULT
                }
            val known = mutableListOf<CalendarCardConfig>()
            for (i in 0 until raw.length()) {
                raw.optJSONObject(i)?.let { obj ->
                    CalendarCardKind.fromKey(obj.optString("kind", "").ifBlank { null })?.let { kind ->
                        val isVisible = if (obj.has("isVisible")) obj.optBoolean("isVisible", true) else true
                        val span = CalendarCardSpan.fromColumns(if (obj.has("span")) obj.optInt("span") else null)
                        known.add(CalendarCardConfig(kind = kind, isVisible = isVisible, span = span ?: kind.defaultSpan))
                    }
                }
            }
            return CalendarPageLayout(reconciled(known))
        }

        /** Deduplicates and appends any kinds missing from [configs] so every
         *  registered card always has exactly one slot. */
        private fun reconciled(configs: List<CalendarCardConfig>): List<CalendarCardConfig> {
            val seen = mutableSetOf<CalendarCardKind>()
            val result = mutableListOf<CalendarCardConfig>()
            for (config in configs) {
                if (seen.add(config.kind)) result.add(config)
            }
            for (kind in CalendarCardKind.entries) {
                if (kind !in seen) result.add(CalendarCardConfig(kind = kind))
            }
            return result
        }
    }
}

// MARK: - Persistence

/**
 * Persists the Calendar panel layout as JSON in preferences DataStore so the
 * arrangement survives across launches. Follows the `QuotaPreferences`
 * pattern: a process-wide singleton with an eager `StateFlow`, so the value
 * stays sticky if the screen is rebuilt.
 */
class CalendarPageLayoutStore private constructor(private val context: Context) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    val layout: StateFlow<CalendarPageLayout> =
        context.dataStore.data
            .map { prefs -> CalendarPageLayout.decode(prefs[KEY_PAGE_LAYOUT_JSON]) }
            .stateIn(scope, SharingStarted.Eagerly, CalendarPageLayout.DEFAULT)

    fun persist(layout: CalendarPageLayout) {
        scope.launch {
            context.dataStore.edit { prefs ->
                prefs[KEY_PAGE_LAYOUT_JSON] = layout.encode()
            }
        }
    }

    companion object {
        private val Context.dataStore by preferencesDataStore("burnbar.calendar.prefs")
        private val KEY_PAGE_LAYOUT_JSON = stringPreferencesKey("calendar_page_layout_json_v1")

        @Volatile private var instance: CalendarPageLayoutStore? = null

        fun get(context: Context): CalendarPageLayoutStore = instance ?: synchronized(this) {
            instance ?: CalendarPageLayoutStore(context.applicationContext).also { instance = it }
        }
    }
}

File: AgentLens/Views/Calendar/CalendarPageLayout.swift
import Foundation

// MARK: - Calendar Card Kind
//
// The registry of every analytics card the Calendar surface can render for
// the current day selection. Mirrors `ChartKind`: each kind carries its own
// editorial metadata, and adding a card = add a case here, prepare its data
// in `CalendarSelectionSnapshot`, and give it a renderer arm in
// `CalendarAnalyticsPanel`.

enum CalendarCardKind: String, Codable, CaseIterable, Identifiable, Sendable {
    case kpis
    case burnOverSelection
    case providerMix
    case modelMix
    case hourOfDayHeatmap
    case projectFocus
    case cacheROI
    case reasoningShare

    var id: String { rawValue }

    var title: String {
        switch self {
        case .kpis: return "Key Numbers"
        case .burnOverSelection: return "Burn Over Selection"
        case .providerMix: return "Provider Mix"
        case .modelMix: return "Model Mix"
        case .hourOfDayHeatmap: return "When You Burn"
        case .projectFocus: return "Project Focus"
        case .cacheROI: return "Cache Savings"
        case .reasoningShare: return "Reasoning Share"
        }
    }

    /// One-line "why it matters" — rendered as the card's footer microcopy.
    var whyItMatters: String {
        switch self {
        case .kpis:
            return "The selection at a glance — spend, volume, sessions, cadence."
        case .burnOverSelection:
            return "Day-by-day spend across the selected days."
        case .providerMix:
            return "Where the money actually goes across providers."
        case .modelMix:
            return "Which models earn their keep — and which quietly dominate."
        case .hourOfDayHeatmap:
            return "Your true working rhythm inside the selection."
        case .projectFocus:
            return "Which projects the selected days actually funded."
        case .cacheROI:
            return "Prompt caching pays rent. This is the receipt."
        case .reasoningShare:
            return "Extended thinking is a hidden line item. Watch its share."
        }
    }

    var systemImage: String {
        switch self {
        case .kpis: return "number.square"
        case .burnOverSelection: return "chart.bar"
        case .providerMix: return "chart.pie"
        case .modelMix: return "cube.transparent"
        case .hourOfDayHeatmap: return "clock.badge"
        case .projectFocus: return "scope"
        case .cacheROI: return "archivebox"
        case .reasoningShare: return "brain"
        }
    }

    /// Grid span in columns (the analytics grid is 3 columns wide; 3 = full
    /// width). Surfaced in the UI as S/M/L sizes.
    var defaultSpan: Int {
        switch self {
        case .kpis, .burnOverSelection, .hourOfDayHeatmap:
            return 3
        default:
            return 1
        }
    }

    var defaultVisible: Bool { true }
}

// MARK: - Card configuration & page layout

/// One card's persisted state: which card, whether shown, and how wide.
struct CalendarCardConfig: Codable, Equatable, Identifiable, Sendable {
    var kind: CalendarCardKind
    var isVisible: Bool
    var span: Int

    var id: String { kind.rawValue }

    init(kind: CalendarCardKind, isVisible: Bool? = nil, span: Int? = nil) {
        self.kind = kind
        self.isVisible = isVisible ?? kind.defaultVisible
        self.span = min(3, max(1, span ?? kind.defaultSpan))
    }
}

/// The full panel layout: an ordered list of card configs, persisted as JSON
/// under `CalendarPageLayout.storageKey`. Same forward-compatible contract as
/// `ChartsPageLayout` — unknown kinds are dropped, missing kinds are appended
/// with defaults, so upgrades never clobber the user's arrangement.
struct CalendarPageLayout: Equatable, Sendable {
    static let storageKey = "calendarPageLayout.v1"

    private(set) var configs: [CalendarCardConfig]

    static var `default`: CalendarPageLayout {
        CalendarPageLayout(configs: CalendarCardKind.allCases.map { CalendarCardConfig(kind: $0) })
    }

    init(configs: [CalendarCardConfig]) {
        self.configs = Self.reconciled(configs)
    }

    /// Cards the grid actually renders, in order.
    var visibleConfigs: [CalendarCardConfig] {
        configs.filter(\.isVisible)
    }

    var hiddenConfigs: [CalendarCardConfig] {
        configs.filter { !$0.isVisible }
    }

    // MARK: Mutations

    /// Moves `kind` so it occupies the position currently held by `target`
    /// (drag-to-reorder lands on a specific card, not an abstract index).
    mutating func move(_ kind: CalendarCardKind, toPositionOf target: CalendarCardKind) {
        guard kind != target,
              let from = configs.firstIndex(where: { $0.kind == kind }),
              let to = configs.firstIndex(where: { $0.kind == target }) else { return }
        let config = configs.remove(at: from)
        configs.insert(config, at: to)
    }

    mutating func setVisible(_ kind: CalendarCardKind, _ visible: Bool) {
        guard let index = configs.firstIndex(where: { $0.kind == kind }) else { return }
        configs[index].isVisible = visible
    }

    mutating func setSpan(_ kind: CalendarCardKind, _ span: Int) {
        guard let index = configs.firstIndex(where: { $0.kind == kind }) else { return }
        configs[index].span = min(3, max(1, span))
    }

    mutating func reset() {
        self = .default
    }

    // MARK: Persistence

    /// JSON round-trip tolerant of unknown kinds. See type comment.
    static func decode(from data: Data) -> CalendarPageLayout {
        struct RawConfig: Codable {
            let kind: String
            let isVisible: Bool?
            let span: Int?
        }
        guard let raw = try? JSONDecoder().decode([RawConfig].self, from: data) else {
            return .default
        }
        let known = raw.compactMap { entry -> CalendarCardConfig? in
            guard let kind = CalendarCardKind(rawValue: entry.kind) else { return nil }
            return CalendarCardConfig(kind: kind, isVisible: entry.isVisible, span: entry.span)
        }
        return CalendarPageLayout(configs: known)
    }

    func encoded() -> Data? {
        try? JSONEncoder().encode(configs)
    }

    /// Deduplicates and appends any kinds missing from `configs` so every
    /// registered card always has exactly one slot.
    private static func reconciled(_ configs: [CalendarCardConfig]) -> [CalendarCardConfig] {
        var seen = Set<CalendarCardKind>()
        var result: [CalendarCardConfig] = []
        for config in configs where !seen.contains(config.kind) {
            seen.insert(config.kind)
            result.append(config)
        }
        for kind in CalendarCardKind.allCases where !seen.contains(kind) {
            result.append(CalendarCardConfig(kind: kind))
        }
        return result
    }
}
